server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java

```diff
@@ -175,4 +177,31 @@ RuntimeField parse(String name, Map<String, Object> node, Mapper.TypeParser.Pars
         }
         return Collections.unmodifiableMap(runtimeFields);
     }
+
+    /**
+     * Collect and return all {@link MappedFieldType} exposed by the provided {@link RuntimeField}s.
+     * Note that validation is performed to make sure that there are no name clashes among the collected runtime fields.
+     * This is because runtime fields with the same name are not accepted as part of the same section.
+     * @param runtimeFields the runtime to extract the mapped field types from
+     * @return the collected mapped field types
+     */
+    static Map<String, MappedFieldType> collectFieldTypes(Collection<RuntimeField> runtimeFields) {
+        return runtimeFields.stream()
+            .flatMap(runtimeField -> {
+                List<String> names = runtimeField.asMappedFieldTypes().stream().map(MappedFieldType::name)
+                    .filter(name -> name.equals(runtimeField.name()) == false
+                        && (name.startsWith(runtimeField.name() + ".") == false
+                        || name.length() > runtimeField.name().length() + 1 == false))
+                    .collect(Collectors.toList());
+                if (names.isEmpty() == false) {
+                    throw new IllegalArgumentException("Found sub-fields with name not belonging to the parent field they are part of "
+                        + names);
```

Comment:
This is interesting, are we sure that we want to enforce this?  We've discussed in the past wanting to have subfields that are aliased out to the root so that they don't appear in their parent object's namespace.  If we *do* enforce this then that's another reason to allow runtime aliases to refer to other runtime fields.

Reply:
I was assuming that we do want to enforce this, otherwise object "foo" can emit sub-fields that don't belong to its namespace. We were trying to apply a similar assertion with the dynamic field type approach but it turned out not to be feasible due to flattened and aliases, I think.

In reality, this will not be possible if we look at how you define sub-fields (yet I think it's good to check to prevent mistakes), because the full name is generated by concatenating the name of the parent and the name of the sub-field:

```
"log" : {
    "type" : "I don't have a name yet"
    "script" : "...."
    "fields" : {
        "client_ip" : {
            "type" : "ip"
        }
    }
}
```

Maybe I should make it an illegal state instead of a user error...

While it is possible to have users define the full path in the sub-fields, that would make the "object" name almost useless and would feel weird. Your conclusion about aliases makes sense to me, although it is very easy to make an alias runtime field with a super small script?

Reply:
I made this an IllegalStateException

Reply:
Using `IllegalStateException` or an assertion here makes sense to me, as it'd indicate a bug in our code (or plugin authors' code). Not relevant to this PR, but I noticed we don't assert this for `FieldMapper#iterator`... maybe we should.